File: python/selene_isru/modules/site_profile.py
from __future__ import annotations

import json
import math
from typing import Any


def _scalar_profile(params: dict[str, Any]) -> dict[str, Any]:
    equatorial = params["site"] == "equatorial"
    shadow = 354 if equatorial else params["polarLongestShadowHours"]
    fraction = 0.5 if equatorial else params["polarIlluminationFraction"]
    day = shadow * fraction / (1 - fraction)
    cycle = day + shadow
    transition = min(1e-6, shadow / 1000)
    return {
        "mode": "scalar",
        "name": "Equatorial lunar day/night cycle" if equatorial else "Scalar polar assumptions",
        "cycleHours": cycle,
        "averageIllumination": fraction,
        "averageReceiverVisibility": 1,
        "averageDeliveredFraction": fraction,
        "longestShadowHours": shadow,
        "longestReceiverOutageHours": shadow,
        "minimumSurfaceTemperatureK": params["Tpsr"],
        "maximumSurfaceTemperatureK": params["Tsurface"],
        "points": [
            {"hour": 0, "illumination": 1, "receiverVisibility": 1, "surfaceTemperatureK": params["Tsurface"]},
            {"hour": day, "illumination": 1, "receiverVisibility": 1, "surfaceTemperatureK": params["Tsurface"]},
            {"hour": day + transition, "illumination": 0, "receiverVisibility": 1, "surfaceTemperatureK": params["Tpsr"]},
            {"hour": cycle, "illumination": 0, "receiverVisibility": 1, "surfaceTemperatureK": params["Tpsr"]},
        ],
    }


def _sample_points(points: list[dict[str, float]], hour: float) -> dict[str, float]:
    cycle = points[-1]["hour"]
    t = ((hour % cycle) + cycle) % cycle
    index = 0
    while index < len(points) - 2 and points[index + 1]["hour"] < t:
        index += 1
    a = points[index]
    b = points[index + 1] if index + 1 < len(points) else a
    span = max(1e-12, b["hour"] - a["hour"])
    fraction = min(1, max(0, (t - a["hour"]) / span))

    def lerp(x: float, y: float) -> float:
        return x + (y - x) * fraction

    return {
        "hour": t,
        "illumination": lerp(a["illumination"], b["illumination"]),
        "receiverVisibility": lerp(a["receiverVisibility"], b["receiverVisibility"]),
        "surfaceTemperatureK": lerp(a["surfaceTemperatureK"], b["surfaceTemperatureK"]),
    }


def _longest_circular_outage(points: list[dict[str, float]], include_visibility: bool) -> float:
    cycle = points[-1]["hour"]
    bins = 1024
    dark: list[bool] = []
    for index in range(bins):
        point = _sample_points(points, ((index + 0.5) / bins) * cycle)
        delivered = point["illumination"] * (point["receiverVisibility"] if include_visibility else 1)
        dark.append(delivered <= 0.05)
    longest = 0
    run = 0
    for index in range(bins * 2):
        if dark[index % bins]:
            run = min(bins, run + 1)
            longest = max(longest, run)
        else:
            run = 0
    return longest / bins * cycle


def _parse_imported(params: dict[str, Any]) -> dict[str, Any] | None:
    payload = params["polarProfileData"]
    if not isinstance(payload, str) or not payload or len(payload) > 100_000:
        return None
    try:
        parsed = json.loads(payload)
    except (json.JSONDecodeError, TypeError):
        return None
    raw_points = parsed.get("points") if isinstance(parsed, dict) else None
    if not isinstance(raw_points, list) or not 2 <= len(raw_points) <= 512:
        return None
    points: list[dict[str, float]] = []
    for raw in raw_points:
        if not isinstance(raw, dict):
            return None
        hour = raw.get("hour")
        illumination = raw.get("illumination")
        visibility = raw.get("receiverVisibility", 1)
        temperature = raw.get("surfaceTemperatureK", params["Tsurface"])
        values = (hour, illumination, visibility, temperature)
        if any(not isinstance(value, (int, float)) or isinstance(value, bool) or not math.isfinite(value) for value in values):
            return None
        if hour < 0 or not 0 <= illumination <= 1 or not 0 <= visibility <= 1 or not 20 <= temperature <= 450:
            return None
        if points and hour <= points[-1]["hour"]:
            return None
        points.append({"hour": float(hour), "illumination": float(illumination), "receiverVisibility": float(visibility), "surfaceTemperatureK": float(temperature)})
    if abs(points[0]["hour"]) > 1e-9 or points[-1]["hour"] <= 0:
        return None
    cycle = points[-1]["hour"]
    illumination_integral = 0.0
    visibility_integral = 0.0
    delivered_integral = 0.0
    for a, b in zip(points, points[1:]):
        dt = b["hour"] - a["hour"]
        illumination_integral += 0.5 * (a["illumination"] + b["illumination"]) * dt
        visibility_integral += 0.5 * (a["receiverVisibility"] + b["receiverVisibility"]) * dt
        delivered_integral += dt / 6 * (
            2 * a["illumination"] * a["receiverVisibility"]
            + a["illumination"] * b["receiverVisibility"]
            + b["illumination"] * a["receiverVisibility"]
            + 2 * b["illumination"] * b["receiverVisibility"]
        )
    name = parsed.get("name")
    return {
        "mode": "profile",
        "name": name.strip()[:80] if isinstance(name, str) and name.strip() else "Imported polar profile",
        "cycleHours": cycle,
        "averageIllumination": illumination_integral / cycle,
        "averageReceiverVisibility": visibility_integral / cycle,
        "averageDeliveredFraction": delivered_integral / cycle,
        "longestShadowHours": _longest_circular_outage(points, False),
        "longestReceiverOutageHours": _longest_circular_outage(points, True),
        "minimumSurfaceTemperatureK": min(point["surfaceTemperatureK"] for point in points),
        "maximumSurfaceTemperatureK": max(point["surfaceTemperatureK"] for point in points),
        "points": points,
    }


def resolve_polar_profile(params: dict[str, Any]) -> dict[str, Any]:
    if params["site"] != "polar" or params["polarProfileMode"] == "scalar":
        return {"profile": _scalar_profile(params), "warnings": []}
    imported = _parse_imported(params)
    if imported is not None:
        return {"profile": imported, "warnings": []}
    return {
        "profile": _scalar_profile(params),
        "warnings": [{"id": "polar-profile-invalid", "severity": "alarm", "module": "power", "message": "The imported polar site profile is invalid; scalar assumptions are active.", "value": 0, "limit": 1}],
    }


def sample_polar_profile(profile: dict[str, Any], hour: float) -> dict[str, float]:
    return _sample_points(profile["points"], hour)

File: packages/engine/src/modules/siteProfile.ts
import type { PolarProfilePoint, PolarProfileSummary, SimParams, Warning } from "../types";

interface ImportedProfile {
  version?: number;
  name?: string;
  points?: Array<{
    hour?: number;
    illumination?: number;
    receiverVisibility?: number;
    surfaceTemperatureK?: number;
  }>;
}

export interface SiteProfileResult {
  profile: PolarProfileSummary;
  warnings: Warning[];
}

function scalarProfile(params: SimParams): PolarProfileSummary {
  const equatorial = params.site === "equatorial";
  const shadow = equatorial ? 354 : params.polarLongestShadowHours;
  const fraction = equatorial ? 0.5 : params.polarIlluminationFraction;
  const day = shadow * fraction / (1 - fraction);
  const cycle = day + shadow;
  const transition = Math.min(1e-6, shadow / 1000);
  return {
    mode: "scalar",
    name: equatorial ? "Equatorial lunar day/night cycle" : "Scalar polar assumptions",
    cycleHours: cycle,
    averageIllumination: fraction,
    averageReceiverVisibility: 1,
    averageDeliveredFraction: fraction,
    longestShadowHours: shadow,
    longestReceiverOutageHours: shadow,
    minimumSurfaceTemperatureK: params.Tpsr,
    maximumSurfaceTemperatureK: params.Tsurface,
    points: [
      { hour: 0, illumination: 1, receiverVisibility: 1, surfaceTemperatureK: params.Tsurface },
      { hour: day, illumination: 1, receiverVisibility: 1, surfaceTemperatureK: params.Tsurface },
      { hour: day + transition, illumination: 0, receiverVisibility: 1, surfaceTemperatureK: params.Tpsr },
      { hour: cycle, illumination: 0, receiverVisibility: 1, surfaceTemperatureK: params.Tpsr }
    ]
  };
}

function finite(value: unknown): value is number {
  return typeof value === "number" && Number.isFinite(value);
}

function parseImported(params: SimParams): PolarProfileSummary | null {
  if (params.polarProfileData.length === 0 || params.polarProfileData.length > 100_000) {
    return null;
  }
  let parsed: ImportedProfile;
  try {
    parsed = JSON.parse(params.polarProfileData) as ImportedProfile;
  } catch {
    return null;
  }
  if (!Array.isArray(parsed.points) || parsed.points.length < 2 || parsed.points.length > 512) {
    return null;
  }
  const points: PolarProfilePoint[] = [];
  for (const raw of parsed.points) {
    if (!finite(raw.hour) || !finite(raw.illumination)) {
      return null;
    }
    const receiverVisibility = raw.receiverVisibility ?? 1;
    const surfaceTemperatureK = raw.surfaceTemperatureK ?? params.Tsurface;
    if (!finite(receiverVisibility) || !finite(surfaceTemperatureK)) {
      return null;
    }
    if (
      raw.hour < 0 || raw.illumination < 0 || raw.illumination > 1 ||
      receiverVisibility < 0 || receiverVisibility > 1 ||
      surfaceTemperatureK < 20 || surfaceTemperatureK > 450
    ) {
      return null;
    }
    if (points.length > 0 && raw.hour <= points[points.length - 1]!.hour) {
      return null;
    }
    points.push({ hour: raw.hour, illumination: raw.illumination, receiverVisibility, surfaceTemperatureK });
  }
  if (Math.abs(points[0]!.hour) > 1e-9 || points.at(-1)!.hour <= 0) {
    return null;
  }

  const cycleHours = points.at(-1)!.hour;
  let illumIntegral = 0;
  let visibilityIntegral = 0;
  let deliveredIntegral = 0;
  for (let index = 0; index < points.length - 1; index += 1) {
    const a = points[index]!;
    const b = points[index + 1]!;
    const dt = b.hour - a.hour;
    illumIntegral += 0.5 * (a.illumination + b.illumination) * dt;
    visibilityIntegral += 0.5 * (a.receiverVisibility + b.receiverVisibility) * dt;
    // Exact integral of the product of two linearly interpolated fractions.
    deliveredIntegral += dt / 6 * (
      2 * a.illumination * a.receiverVisibility +
      a.illumination * b.receiverVisibility +
      b.illumination * a.receiverVisibility +
      2 * b.illumination * b.receiverVisibility
    );
  }
  const longestShadowHours = longestCircularOutage(points, false);
  const longestReceiverOutageHours = longestCircularOutage(points, true);
  return {
    mode: "profile",
    name: typeof parsed.name === "string" && parsed.name.trim().length > 0
      ? parsed.name.trim().slice(0, 80)
      : "Imported polar profile",
    cycleHours,
    averageIllumination: illumIntegral / cycleHours,
    averageReceiverVisibility: visibilityIntegral / cycleHours,
    averageDeliveredFraction: deliveredIntegral / cycleHours,
    longestShadowHours,
    longestReceiverOutageHours,
    minimumSurfaceTemperatureK: Math.min(...points.map((point) => point.surfaceTemperatureK)),
    maximumSurfaceTemperatureK: Math.max(...points.map((point) => point.surfaceTemperatureK)),
    points
  };
}

function longestCircularOutage(points: PolarProfilePoint[], includeVisibility: boolean): number {
  const cycle = points.at(-1)!.hour;
  const bins = 1024;
  const dark: boolean[] = [];
  for (let index = 0; index < bins; index += 1) {
    const point = samplePoints(points, ((index + 0.5) / bins) * cycle);
    const delivered = point.illumination * (includeVisibility ? point.receiverVisibility : 1);
    dark.push(delivered <= 0.05);
  }
  let longest = 0;
  let run = 0;
  for (let index = 0; index < bins * 2; index += 1) {
    if (dark[index % bins]!) {
      run = Math.min(bins, run + 1);
      longest = Math.max(longest, run);
    } else {
      run = 0;
    }
  }
  return (longest / bins) * cycle;
}

function samplePoints(points: PolarProfilePoint[], hour: number): PolarProfilePoint {
  const cycle = points.at(-1)!.hour;
  const t = ((hour % cycle) + cycle) % cycle;
  let index = 0;
  while (index < points.length - 2 && points[index + 1]!.hour < t) {
    index += 1;
  }
  const a = points[index]!;
  const b = points[index + 1] ?? a;
  const span = Math.max(1e-12, b.hour - a.hour);
  const fraction = Math.min(1, Math.max(0, (t - a.hour) / span));
  const lerp = (x: number, y: number): number => x + (y - x) * fraction;
  return {
    hour: t,
    illumination: lerp(a.illumination, b.illumination),
    receiverVisibility: lerp(a.receiverVisibility, b.receiverVisibility),
    surfaceTemperatureK: lerp(a.surfaceTemperatureK, b.surfaceTemperatureK)
  };
}

export function resolvePolarProfile(params: SimParams): SiteProfileResult {
  if (params.site !== "polar" || params.polarProfileMode === "scalar") {
    return { profile: scalarProfile(params), warnings: [] };
  }
  const imported = parseImported(params);
  if (imported !== null) {
    return { profile: imported, warnings: [] };
  }
  return {
    profile: scalarProfile(params),
    warnings: [{
      id: "polar-profile-invalid",
      severity: "alarm",
      module: "power",
      message: "The imported polar site profile is invalid; scalar assumptions are active.",
      value: 0,
      limit: 1
    }]
  };
}

export function samplePolarProfile(profile: PolarProfileSummary, hour: number): PolarProfilePoint {
  return samplePoints(profile.points, hour);
}
